frontend/server/src/main/java/org/pytorch/serve/util/GPUManager.java

```diff
@@ -0,0 +1,165 @@
+package org.pytorch.serve.util;
+
+import java.util.Map;
+import java.util.HashMap;
+import java.util.TreeMap;
+import java.util.Iterator;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.nio.charset.StandardCharsets;
+import java.io.BufferedReader;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class GPUManager {
+
+    private static final Logger logger = LoggerFactory.getLogger(GPUManager.class);
+    private static final int nFailureHistory = 100;
+
+    private static GPUManager instance;
+
+    private final int nGPUs;
+    private final int minFreeMemory;
+    private final float maxShareFailures;
+
+    private AtomicInteger[] freeMemory;
+    private ConcurrentHashMap<String, Integer> workerIds;
```

Comment:
As we are using synchronized access methods, we won't need these fields to be able to deal with concurrency - so we can also use `int[]` (or better `ArrayList<Integer>`) and ` HashMap<String, Integer>`.